File: tools/check_fdother.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

int main() {
    FILE* fp = fopen("D:/workspace/fd2_dat_freebuff/bin/FDOTHER.DAT", "rb");
    if (!fp) {
        printf("Cannot open FDOTHER.DAT\n");
        return 1;
    }
    
    /* 读取前200字节 */
    uint8_t header[200];
    fread(header, 1, 200, fp);
    fclose(fp);
    
    printf("FDOTHER.DAT header (first 200 bytes):\n");
    for (int i = 0; i < 200; i++) {
        if (i % 16 == 0) printf("\n%04X: ", i);
        printf("%02X ", header[i]);
    }
    printf("\n\n");
    
    /* 解析偏移表 */
    printf("Offset table analysis:\n");
    printf("Header bytes 0-5: ");
    for (int i = 0; i < 6; i++) {
        printf("%02X ", header[i]);
    }
    printf("\n");
    
    uint16_t header_val = *(uint16_t*)(header + 0);
    uint32_t offset_table_start = *(uint32_t*)(header + 2);
    
    printf("Header value (2 bytes): %u (0x%X)\n", header_val, header_val);
    printf("Offset table start (4 bytes): %u (0x%X)\n", offset_table_start, offset_table_start);
    
    /* 尝试从不同位置读取偏移表 */
    printf("\nOffsets from position 6:\n");
    for (int i = 0; i <= 20; i++) {
        uint32_t offset = *(uint32_t*)(header + 6 + 4*i);
        printf("  [%2d] offset = %u (0x%X)\n", i, offset, offset);
    }
    
    /* 验证索引1、20、3、5的资源数据 */
    printf("\nResource data verification:\n");
    
    /* 索引1 */
    uint32_t off1 = *(uint32_t*)(header + 6 + 4*1);
    printf("\nIndex 1: offset=%u\n", off1);
    if (off1 < 200) {
        printf("  Data at offset: ");
        for (int i = 0; i < 8; i++) {
            printf("%02X ", header[off1 + i]);
        }
        printf("\n");
        int w = header[off1 + 0] | (header[off1 + 1] << 8);
        int h = header[off1 + 2] | (header[off1 + 3] << 8);
        printf("  Dimensions: %dx%d\n", w, h);
    }
    
    /* 索引20 */
    uint32_t off20 = *(uint32_t*)(header + 6 + 4*20);
    printf("\nIndex 20: offset=%u\n", off20);
    if (off20 < 200) {
        printf("  Data at offset: ");
        for (int i = 0; i < 8; i++) {
            printf("%02X ", header[off20 + i]);
        }
        printf("\n");
        int w = header[off20 + 0] | (header[off20 + 1] << 8);
        int h = header[off20 + 2] | (header[off20 + 3] << 8);
        printf("  Dimensions: %dx%d\n", w, h);
    }
    
    return 0;
}

File: tools/check_fdother.py
import struct
import os

dat_path = 'game/FDOTHER.DAT'
if not os.path.exists(dat_path):
    print(f"Error: {dat_path} not found")
    exit(1)

with open(dat_path, 'rb') as f:
    magic = f.read(6)
    print(f'Magic: {magic}')
    
    count = struct.unpack('<I', f.read(4))[0]
    print(f'Resource count: {count}')
    
    offsets = []
    for i in range(count):
        offsets.append(struct.unpack('<I', f.read(4))[0])
    
    print(f'\nOffsets[0..19]:')
    for i in range(min(20, count)):
        print(f'  [{i}] = {offsets[i]}')
    
    print(f'\nResource sizes:')
    for i in range(min(15, count)):
        start = offsets[i]
        end = offsets[i+1] if i+1 < count else os.path.getsize(dat_path)
        size = end - start
        print(f'  Resource {i}: offset={start}, size={size}')
    
    print(f'\nChecking resource #7 (menu resource set):')
    start7 = offsets[7]
    end7 = offsets[8] if 8 < count else os.path.getsize(dat_path)
    size7 = end7 - start7
    print(f'  Resource 7: offset={start7}, size={size7}')
    
    # Read resource #7 header to check if it's a container
    f.seek(start7)
    header = f.read(16)
    print(f'  Header bytes: {header.hex()}')
    
    # Check first few bytes for possible dimensions
    if size7 > 4:
        f.seek(start7)
        w = struct.unpack('<H', f.read(2))[0]
        h = struct.unpack('<H', f.read(2))[0]
        print(f'  Possible dimensions: {w}x{h}')
